test/util/rotate.go

```diff
@@ -0,0 +1,171 @@
+// trusted fashion.
+func CheckRotation(t *testing.T, dnsName string, oldCertPEM, oldKeyPEM, oldBundlePEM, newCertPEM, newKeyPEM, newBundlePEM []byte) {
+	testCases := map[string]struct {
+		certPEM   []byte
+		keyPEM    []byte
+		bundlePEM []byte
+	}{
+		"Pre-rotation": {
+			certPEM:   oldCertPEM,
+			keyPEM:    oldKeyPEM,
+			bundlePEM: oldBundlePEM,
+		},
+		"Server rotated": {
+			certPEM:   newCertPEM,
+			keyPEM:    newKeyPEM,
+			bundlePEM: oldBundlePEM,
+		},
+		"Client refreshed": {
+			certPEM:   oldCertPEM,
+			keyPEM:    oldKeyPEM,
+			bundlePEM: newBundlePEM,
+		},
+		"Server rotated and client refreshed": {
+			certPEM:   newCertPEM,
+			keyPEM:    newKeyPEM,
+			bundlePEM: newBundlePEM,
+		},
+	}
+	for testName, tc := range testCases {
+		t.Run(testName, func(t *testing.T) {
+			checkClientTrust(t, testName, dnsName, tc.certPEM, tc.keyPEM, tc.bundlePEM)
+		})
+	}
+}
+
+// checkClientTrust verifies that a server configured with the provided cert and key will be
+// trusted by a client with the given bundle.
+func checkClientTrust(t *testing.T, testName, dnsName string, certPEM, keyPEM, bundlePEM []byte) {
+	// Emulate how a service will consume the serving cert by writing
+	// the cert and key to disk.
+	certFile, err := ioutil.TempFile("", v1.TLSCertKey)
+	if err != nil {
+		t.Fatalf("error creating tmpfile for cert: %v", err)
+
+	}
+	defer func() {
+		err := os.Remove(certFile.Name())
+		if err != nil {
+			t.Errorf("Failed to remove file %v", err)
+		}
+	}()
+	_, err = certFile.Write(certPEM)
+	if err != nil {
+		t.Fatalf("Error writing cert to disk: %v", err)
+	}
+
+	keyFile, err := ioutil.TempFile("", v1.TLSPrivateKeyKey)
+	if err != nil {
+		t.Fatalf("error creating tmpfile for key: %v", err)
+
+	}
+	defer func() {
+		err := os.Remove(keyFile.Name())
+		if err != nil {
+			t.Errorf("Failed to remove file %v", err)
+		}
+	}()
+	_, err = keyFile.Write(keyPEM)
+	if err != nil {
+		t.Fatalf("Error writing key to disk: %v", err)
+	}
+
+	// The need to listen on a random port precludes the use of ListenAndServeTLS since that
+	// method provides no way to determine the port that the server ends up listening
+	// on. Creating a listener and using ServeTLS ensures a random port will be allocated
+	// (by specifying ':0') and that the resulting port is discoverable via the listener's
+	// Addr() method.
+	listenerAddress := "127.0.0.1:0"
+	ln, err := net.Listen("tcp", listenerAddress)
+	if err != nil {
+		t.Fatalf("net.Listen: %v", err)
+	}
+	defer func() {
+		_ = ln.Close()
+	}()
+	serverAddress := ln.Addr().String()
+	serverPort := serverAddress[strings.LastIndex(serverAddress, ":")+1:]
+
+	srv := http.Server{}
+	// Start a server configured with the cert and key
+	go func() {
+		if err := srv.ServeTLS(ln, certFile.Name(), keyFile.Name()); err != nil && err != http.ErrServerClosed {
+			t.Fatalf("ServeTLS failed: %v", err)
+		}
+	}()
+	defer func() {
+		err = srv.Close()
+		if err != nil {
```

Comment:
nit: `if err := srv.Close(); err != nil`, unless you meant to assign to the outer scope error

Reply:
Done